frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -0,0 +1,72 @@
+package com.google.tripmeout.frontend.storage;
+ 
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.storage.PlaceVisitStorage;
+import com.google.tripmeout.frontend.error.PlaceVisitAlreadyExistsException;
+import com.google.tripmeout.frontend.error.PlaceVisitNotFoundException;
+import com.google.common.collect.Tables;
+import com.google.common.collect.Table;
+import com.google.common.collect.HashBasedTable;
+import java.util.Map;
+import java.util.List;
+import java.util.Collection;
+import java.util.ArrayList;
+
+public class InMemoryPlaceVisitStorage implements PlaceVisitStorage {
+  // <placeId, tripId, PlaceVisitModel> 
+  Table<String, String, PlaceVisitModel> storage = Tables.synchronizedTable(HashBasedTable.create());
+ 
+  @Override
+  public void addPlaceVisit(PlaceVisitModel placeVisit) throws PlaceVisitAlreadyExistsException {
```

Comment:
This isn't thread safe. If two threads call this method at the same time they can override each-others data and have neither one will get a `PlaceVisitAlreadyExistsException`.

Maybe it was a mistake to tell you to use `Tables.synchronizedTable` because it doesn't look like it has a good conditional update method, meaning you will need to do some lock management yourself.

Reply:
would it be fine to switch to a ConcurrentHashMap then? I saw that it is thread-safe but not necessarily synchronized

Reply:
yeah that would be fine. The real difference is that you want something with a [compute](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#compute-K-java.util.function.BiFunction-)-like functionality which it doesn't seem like Tables have (even though it seems pretty suited to the rest of this work).

Reply:
I might be missing something here (or the code has changed since the initial comment as shown): why isn't `synchronized (storage)` sufficient? That work as long as _all_ accesses to the table are done in the same fashion (some changes would need to be made below). Either way, I would drop the synchronized table and use a regular table (or perhaps a map or two).